File: goldencheck/agent/handoff.py
"""Pipeline handoff module — generates a structured handoff dict for downstream consumers."""
from __future__ import annotations

import hashlib
from collections import defaultdict
from dataclasses import asdict
from datetime import datetime, timezone
from pathlib import Path

from goldencheck import __version__
from goldencheck.models.finding import Finding, Severity
from goldencheck.models.profile import DatasetProfile

__all__ = ["generate_handoff", "findings_to_fbc"]


def findings_to_fbc(findings: list[Finding]) -> dict[str, dict[str, int]]:
    """Convert a list of findings to a findings-by-column dict for health_score.

    Returns ``{col_name: {"errors": N, "warnings": N}}`` with counts aggregated
    per column.
    """
    fbc: dict[str, dict[str, int]] = defaultdict(lambda: {"errors": 0, "warnings": 0})
    for f in findings:
        if f.severity == Severity.ERROR:
            fbc[f.column]["errors"] += 1
        elif f.severity == Severity.WARNING:
            fbc[f.column]["warnings"] += 1
    return dict(fbc)


def _compute_file_hash(file_path: str) -> str:
    """Return ``sha256:<hex>`` digest of the file at *file_path*."""
    h = hashlib.sha256()
    with Path(file_path).open("rb") as fh:
        for chunk in iter(lambda: fh.read(1 << 16), b""):
            h.update(chunk)
    return f"sha256:{h.hexdigest()}"


def _attestation(
    findings: list[Finding],
    review_pending: int,
    dismissed: int,  # noqa: ARG001 — kept for signature symmetry
) -> str:
    """Derive the attestation label from findings and review state."""
    has_errors = any(f.severity == Severity.ERROR for f in findings)
    has_warnings = any(f.severity == Severity.WARNING for f in findings)

    if has_errors:
        return "FAIL"
    if review_pending > 0:
        return "REVIEW_REQUIRED"
    if has_warnings:
        return "PASS_WITH_WARNINGS"
    return "PASS"


def _build_columns(
    profile: DatasetProfile,
    findings: list[Finding],
    column_types: dict[str, str] | None,
) -> dict[str, dict]:
    """Build the per-column summary dict."""
    # Group findings by column
    issues_by_col: dict[str, list[dict]] = defaultdict(list)
    for f in findings:
        issues_by_col[f.column].append({
            "check": f.check,
            "severity": f.severity.name,
            "message": f.message,
            "confidence": f.confidence,
            "affected_rows": f.affected_rows,
        })

    columns: dict[str, dict] = {}
    for cp in profile.columns:
        col_type = (column_types or {}).get(cp.name, cp.inferred_type)
        columns[cp.name] = {
            "type": col_type,
            "null_pct": round(cp.null_pct, 4),
            "unique_pct": round(cp.unique_pct, 4),
            "issues": issues_by_col.get(cp.name, []),
        }
    return columns


def generate_handoff(
    file_path: str,
    findings: list[Finding],
    profile: DatasetProfile,
    pinned_rules: list[dict],
    review_pending: int,
    dismissed: int,
    job_name: str,
    column_types: dict[str, str] | None = None,
) -> dict:
    """Generate a structured handoff dict summarising a GoldenCheck scan.

    The returned dict is JSON-serialisable and follows ``schema_version: 1``.
    """
    # Counts
    error_count = sum(1 for f in findings if f.severity == Severity.ERROR)
    warning_count = sum(1 for f in findings if f.severity == Severity.WARNING)

    # Health score via profile helper
    fbc = findings_to_fbc(findings)
    grade, score = profile.health_score(findings_by_column=fbc)

    # Unresolved: medium-confidence findings (not yet pinned or dismissed)
    unresolved = [
        asdict(f)
        for f in findings
        if 0.5 <= f.confidence < 0.8
    ]

    return {
        "schema_version": 1,
        "source_tool": "goldencheck",
        "source_version": __version__,
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "job_name": job_name,
        "file_path": file_path,
        "file_hash": _compute_file_hash(file_path),
        "row_count": profile.row_count,
        "column_count": profile.column_count,
        "health": {"grade": grade, "score": score},
        "summary": {
            "errors": error_count,
            "warnings": warning_count,
            "pinned_rules": len(pinned_rules),
            "review_pending": review_pending,
            "dismissed": dismissed,
        },
        "columns": _build_columns(profile, findings, column_types),
        "pinned_rules": pinned_rules,
        "unresolved_findings": unresolved,
        "attestation": _attestation(findings, review_pending, dismissed),
    }

File: packages/goldencheck-js/src/core/agent/handoff.ts
/**
 * Pipeline handoff module — generates a structured handoff dict for downstream consumers.
 * Port of goldencheck/agent/handoff.py.
 * Edge-safe: no Node.js dependencies.
 */

import type { Finding, DatasetProfile, ColumnProfile } from "../types.js";
import { Severity, severityLabel, healthScore } from "../types.js";

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/**
 * Convert a list of findings to a findings-by-column dict for healthScore().
 */
function findingsToFbc(
  findings: readonly Finding[],
): Record<string, { errors: number; warnings: number }> {
  const fbc: Record<string, { errors: number; warnings: number }> = {};
  for (const f of findings) {
    if (!fbc[f.column]) {
      fbc[f.column] = { errors: 0, warnings: 0 };
    }
    if (f.severity === Severity.ERROR) {
      fbc[f.column]!.errors += 1;
    } else if (f.severity === Severity.WARNING) {
      fbc[f.column]!.warnings += 1;
    }
  }
  return fbc;
}

/**
 * Derive the attestation label from findings and review state.
 */
function attestation(
  findings: readonly Finding[],
  reviewPending: number,
): string {
  const hasErrors = findings.some((f) => f.severity === Severity.ERROR);
  const hasWarnings = findings.some((f) => f.severity === Severity.WARNING);

  if (hasErrors) return "FAIL";
  if (reviewPending > 0) return "REVIEW_REQUIRED";
  if (hasWarnings) return "PASS_WITH_WARNINGS";
  return "PASS";
}

/**
 * Build the per-column summary dict.
 */
function buildColumns(
  profile: DatasetProfile,
  findings: readonly Finding[],
  columnTypes?: Readonly<Record<string, string>> | null,
): Record<string, Record<string, unknown>> {
  // Group findings by column
  const issuesByCol: Record<string, object[]> = {};
  for (const f of findings) {
    if (!issuesByCol[f.column]) {
      issuesByCol[f.column] = [];
    }
    issuesByCol[f.column]!.push({
      check: f.check,
      severity: severityLabel(f.severity),
      message: f.message,
      confidence: f.confidence,
      affected_rows: f.affectedRows,
    });
  }

  const columns: Record<string, Record<string, unknown>> = {};
  for (const cp of profile.columns) {
    const colType = columnTypes?.[cp.name] ?? cp.inferredType;
    columns[cp.name] = {
      type: colType,
      null_pct: Math.round(cp.nullPct * 10000) / 10000,
      unique_pct: Math.round(cp.uniquePct * 10000) / 10000,
      issues: issuesByCol[cp.name] ?? [],
    };
  }
  return columns;
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

export interface GenerateHandoffOptions {
  filePath: string;
  findings: readonly Finding[];
  profile: DatasetProfile;
  pinnedRules: readonly object[];
  reviewPending: number;
  dismissed: number;
  jobName: string;
  columnTypes?: Readonly<Record<string, string>> | null;
}

/**
 * Generate a structured handoff dict summarising a GoldenCheck scan.
 *
 * The returned object is JSON-serialisable and follows schema_version: 1.
 *
 * Attestation logic:
 * - FAIL: any ERROR finding
 * - REVIEW_REQUIRED: reviewPending > 0
 * - PASS_WITH_WARNINGS: any WARNING
 * - PASS: clean
 */
export function generateHandoff(opts: GenerateHandoffOptions): object {
  const {
    filePath,
    findings,
    profile,
    pinnedRules,
    reviewPending,
    dismissed,
    jobName,
    columnTypes,
  } = opts;

  // Counts
  const errorCount = findings.filter((f) => f.severity === Severity.ERROR).length;
  const warningCount = findings.filter((f) => f.severity === Severity.WARNING).length;

  // Health score
  const fbc = findingsToFbc(findings);
  const health = healthScore(fbc);

  // Unresolved: medium-confidence findings (not yet pinned or dismissed)
  const unresolvedFindings = findings
    .filter((f) => f.confidence >= 0.5 && f.confidence < 0.8)
    .map((f) => ({
      severity: severityLabel(f.severity),
      column: f.column,
      check: f.check,
      message: f.message,
      confidence: f.confidence,
      affectedRows: f.affectedRows,
      sampleValues: [...f.sampleValues],
      suggestion: f.suggestion,
      pinned: f.pinned,
      source: f.source,
      metadata: { ...f.metadata },
    }));

  return {
    schema_version: 1,
    source_tool: "goldencheck",
    timestamp: new Date().toISOString(),
    job_name: jobName,
    file_path: filePath,
    row_count: profile.rowCount,
    column_count: profile.columnCount,
    health: { grade: health.grade, score: health.points },
    summary: {
      errors: errorCount,
      warnings: warningCount,
      pinned_rules: pinnedRules.length,
      review_pending: reviewPending,
      dismissed,
    },
    columns: buildColumns(profile, findings, columnTypes),
    pinned_rules: pinnedRules,
    unresolved_findings: unresolvedFindings,
    attestation: attestation(findings, reviewPending),
  };
}
